pkg/mev/send_bundle.go

```diff
@@ -0,0 +1,174 @@
+package mev
+
+import (
+	"bytes"
+	"context"
+	"crypto/ecdsa"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+
+	"github.com/ethereum/go-ethereum/accounts"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+const (
+	SendBundleID              = 1
+	SendBundleBloxrouteMethod = "blxr_simulate_bundle"
+	SendBundleMethod          = "eth_sendBundle"
+)
+
+func SendBundle( // nolint: cyclop
+	ctx context.Context, c *http.Client, endpoint string, req SendBundleRequest, options ...SendBundleOptions,
+) (SendBundleResponse, error) {
+	var opts sendBundleOpts
+	for _, fn := range options {
+		if fn == nil {
+			continue
+		}
+		fn(&opts)
+	}
+
+	reqBody, err := json.Marshal(req)
+	if err != nil {
+		return SendBundleResponse{}, fmt.Errorf("marshal json error: %w", err)
+	}
+	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, bytes.NewBuffer(reqBody))
+	if err != nil {
+		return SendBundleResponse{}, fmt.Errorf("new http request error: %w", err)
+	}
+	if opts.flashbotSignKey != nil {
+		if err := signFlashbotRequest(opts.flashbotSignKey, httpReq, reqBody); err != nil {
+			return SendBundleResponse{}, fmt.Errorf("sign flashbot request error: %w", err)
+		}
+	}
+	httpReq.Header.Add("Content-Type", "application/json")
+	httpReq.Header.Add("Accept", "application/json")
+
+	httpResp, err := c.Do(httpReq)
+	if err != nil {
+		return SendBundleResponse{}, fmt.Errorf("do request error: %w", err)
+	}
+	defer httpResp.Body.Close()
+
+	respBody, err := io.ReadAll(httpResp.Body)
+	if err != nil {
+		return SendBundleResponse{}, fmt.Errorf("read response error: %w", err)
+	}
+
+	if httpResp.StatusCode != http.StatusOK {
+		return SendBundleResponse{}, fmt.Errorf("not OK status, status: [%d], data: [%s]",
+			httpResp.StatusCode, string(respBody))
+	}
+
+	var resp SendBundleResponse
+	if err := json.Unmarshal(respBody, &resp); err != nil {
+		return SendBundleResponse{}, fmt.Errorf("unmarshal response error: %w, data: [%s]", err, string(respBody))
+	}
+
+	return resp, nil
+}
+
+func signFlashbotRequest(key *ecdsa.PrivateKey, request *http.Request, body []byte) error {
+	hashed := crypto.Keccak256Hash(body).Hex()
+	signature, err := crypto.Sign(accounts.TextHash([]byte(hashed)), key)
+	if err != nil {
+		return fmt.Errorf("sign crypto error: %w", err)
+	}
```

Comment:
We can move this code snippet to a function for reusability, for example: other builders accept the same signature as flashbots builder.

Reply:
updated in: https://github.com/KyberNetwork/tradinglib/pull/16/commits/267c4e651fe1aa1cbd2e1f79bde134adaa86d511